src/NHibernate/Async/AdoNet/SqlClientBatchingBatcher.cs

```diff
@@ -0,0 +1,98 @@
+﻿//------------------------------------------------------------------------------
+// <auto-generated>
+//     This code was generated by AsyncGenerator.
+//
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+//------------------------------------------------------------------------------
+
+
+using System;
+using System.Data.Common;
+using System.Text;
+using NHibernate.AdoNet.Util;
+using NHibernate.Exceptions;
+using NHibernate.Util;
+
+namespace NHibernate.AdoNet
+{
+	using System.Threading.Tasks;
+	using System.Threading;
+	/// <content>
+	/// Contains generated async methods
+	/// </content>
+	public partial class SqlClientBatchingBatcher : AbstractBatcher
+	{
+
+		public override Task AddToBatchAsync(IExpectation expectation, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			try
+			{
+				_totalExpectedRowsAffected += expectation.ExpectedRowCount;
+				var batchUpdate = CurrentCommand;
+				Driver.AdjustCommand(batchUpdate);
+				string lineWithParameters = null;
+				var sqlStatementLogger = Factory.Settings.SqlStatementLogger;
+				if (sqlStatementLogger.IsDebugEnabled || Log.IsDebugEnabled)
+				{
+					lineWithParameters = sqlStatementLogger.GetCommandLineWithParameters(batchUpdate);
+					var formatStyle = sqlStatementLogger.DetermineActualStyle(FormatStyle.Basic);
+					lineWithParameters = formatStyle.Formatter.Format(lineWithParameters);
+					_currentBatchCommandsLog.Append("command ")
+					.Append(_currentBatch.CountOfCommands)
+					.Append(":")
+					.AppendLine(lineWithParameters);
+				}
+				if (Log.IsDebugEnabled)
+				{
+					Log.Debug("Adding to batch:" + lineWithParameters);
+				}
+				_currentBatch.Append((System.Data.SqlClient.SqlCommand) batchUpdate);
+
+				if (_currentBatch.CountOfCommands >= _batchSize)
+				{
+					return ExecuteBatchWithTimingAsync(batchUpdate, cancellationToken);
+				}
+				return Task.CompletedTask;
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<object>(ex);
+			}
+		}
+
+		protected override async Task DoExecuteBatchAsync(DbCommand ps, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			Log.DebugFormat("Executing batch");
+			await (CheckReadersAsync(cancellationToken)).ConfigureAwait(false);
+			await (PrepareAsync(_currentBatch.BatchCommand, cancellationToken)).ConfigureAwait(false);
+			if (Factory.Settings.SqlStatementLogger.IsDebugEnabled)
+			{
+				Factory.Settings.SqlStatementLogger.LogBatchCommand(_currentBatchCommandsLog.ToString());
+				_currentBatchCommandsLog = new StringBuilder().AppendLine("Batch commands:");
+			}
+
+			int rowsAffected;
+			try
+			{
+				rowsAffected = _currentBatch.ExecuteNonQuery();
```

Comment:
IMO in the future we will have to write our own batchers that would support async operations especially for `SqlClient` as the `SqlCommandSet` will not be available in .NET Core ([current source](https://github.com/dotnet/corefx/tree/master/src/System.Data.SqlClient/src/System/Data/SqlClient)).